File: kotlin/src/main/kotlin/com/edgeveda/sdk/NativeErrorCode.kt
package com.edgeveda.sdk

/**
 * Maps C core / JNI integer error codes to Kotlin [EdgeVedaException] subclasses.
 *
 * The native C engine returns integer error codes through the JNI boundary.
 * This enum provides a type-safe mapping from those codes to rich Kotlin exceptions.
 * Codes match `ev_error_t` in `edge_veda.h` (all errors are negative; 0 = success).
 *
 * ## Usage
 * ```kotlin
 * val code = NativeErrorCode.fromCode(resultCode)
 * code.throwIfError("Failed during model load")
 * ```
 *
 * ## Error Code Table
 * | Code | Name                 | Description                               |
 * |------|----------------------|-------------------------------------------|
 * |  0   | OK                   | Success (no error)                        |
 * | -1   | INVALID_PARAM        | Invalid parameter value                   |
 * | -2   | OUT_OF_MEMORY        | System ran out of memory                  |
 * | -3   | MODEL_LOAD_FAILED    | Model failed to load                      |
 * | -4   | BACKEND_INIT_FAILED  | Backend (Vulkan/CPU) failed to initialise |
 * | -5   | INFERENCE_FAILED     | Text generation / inference failed        |
 * | -6   | CONTEXT_INVALID      | Context invalid or prompt exceeds window  |
 * | -7   | STREAM_ENDED         | Normal end-of-stream (not an error)       |
 * | -8   | NOT_IMPLEMENTED      | Feature not implemented                   |
 * | -9   | MEMORY_LIMIT_EXCEEDED| Configured memory limit exceeded          |
 * | -10  | UNSUPPORTED_BACKEND  | Backend not supported on this device      |
 * | -999 | UNKNOWN              | Unknown or unmapped error code            |
 */
enum class NativeErrorCode(val code: Int) {
    /** Operation completed successfully (no error) */
    OK(0),

    /** An invalid parameter was passed to the engine */
    INVALID_PARAM(-1),

    /** System ran out of memory during operation */
    OUT_OF_MEMORY(-2),

    /** Model failed to load (corrupt file, unsupported format, etc.) */
    MODEL_LOAD_FAILED(-3),

    /** Backend (Vulkan/CPU) failed to initialise */
    BACKEND_INIT_FAILED(-4),

    /** Text generation / inference failed mid-operation */
    INFERENCE_FAILED(-5),

    /** Context is invalid or prompt exceeded the context window */
    CONTEXT_INVALID(-6),

    /** Normal end-of-stream signal — not an error */
    STREAM_ENDED(-7),

    /** Feature is not yet implemented in the native engine */
    NOT_IMPLEMENTED(-8),

    /** Configured memory limit was exceeded */
    MEMORY_LIMIT_EXCEEDED(-9),

    /** Backend is not supported on this device */
    UNSUPPORTED_BACKEND(-10),

    /** Unknown or unmapped error code */
    UNKNOWN(-999);

    /**
     * Convert this native error code to an [EdgeVedaException].
     *
     * Returns `null` for [OK] and [STREAM_ENDED] since neither represents an error.
     *
     * @param context Optional additional context string for the error message.
     * @return The corresponding [EdgeVedaException], or `null` if the code is [OK]/[STREAM_ENDED].
     */
    fun toException(context: String? = null): EdgeVedaException? {
        val msg = context ?: "Native engine error"

        return when (this) {
            OK -> null
            INVALID_PARAM -> EdgeVedaException.InvalidConfiguration(msg)
            OUT_OF_MEMORY -> EdgeVedaException.OutOfMemoryError(msg)
            MODEL_LOAD_FAILED -> EdgeVedaException.ModelLoadError(msg)
            BACKEND_INIT_FAILED -> EdgeVedaException.ModelLoadError("Backend initialization failed: $msg")
            INFERENCE_FAILED -> EdgeVedaException.GenerationError(msg)
            CONTEXT_INVALID -> EdgeVedaException.ContextOverflowError(msg)
            STREAM_ENDED -> null
            NOT_IMPLEMENTED -> EdgeVedaException.UnsupportedOperationError(msg)
            MEMORY_LIMIT_EXCEEDED -> EdgeVedaException.OutOfMemoryError("Memory limit exceeded: $msg")
            UNSUPPORTED_BACKEND -> EdgeVedaException.InvalidConfiguration("Unsupported backend: $msg")
            UNKNOWN -> EdgeVedaException.NativeError(msg)
        }
    }

    /**
     * Throw the corresponding [EdgeVedaException] if this code is not [OK].
     *
     * @param context Optional context string for the error message.
     * @throws EdgeVedaException if the code represents an error.
     */
    fun throwIfError(context: String? = null) {
        toException(context)?.let { throw it }
    }

    companion object {
        private val codeMap = entries.associateBy { it.code }

        /**
         * Look up a [NativeErrorCode] from a raw integer code.
         *
         * Returns [UNKNOWN] for unmapped values.
         *
         * @param code The raw integer error code from JNI.
         * @return The corresponding [NativeErrorCode].
         */
        fun fromCode(code: Int): NativeErrorCode {
            return codeMap[code] ?: UNKNOWN
        }
    }
}

File: swift/Sources/EdgeVeda/NativeErrorCode.swift
import Foundation

/// Maps C core `ev_error_*` integer codes to Swift `EdgeVedaError` cases.
///
/// The native C engine returns integer error codes through the FFI boundary.
/// This enum provides a type-safe mapping from those codes to rich Swift errors.
///
/// Raw values match the `ev_error_t` enum defined in `edge_veda.h` exactly.
/// Previous versions used positive codes (1–7) which never matched the C API's
/// negative values, causing all errors to map to `.unknown`.
///
/// ## Usage
/// ```swift
/// let code = NativeErrorCode(rawValue: resultCode) ?? .unknown
/// if let error = code.toEdgeVedaError() {
///     throw error
/// }
/// ```
public enum NativeErrorCode: Int32, Sendable {
    /// Operation completed successfully (no error). EV_SUCCESS = 0
    case ok = 0

    /// Invalid parameter provided. EV_ERROR_INVALID_PARAM = -1
    case invalidParameter = -1

    /// Out of memory. EV_ERROR_OUT_OF_MEMORY = -2
    case outOfMemory = -2

    /// Failed to load model. EV_ERROR_MODEL_LOAD_FAILED = -3
    case modelLoadFailed = -3

    /// Failed to initialize backend. EV_ERROR_BACKEND_INIT_FAILED = -4
    case backendInitFailed = -4

    /// Inference operation failed. EV_ERROR_INFERENCE_FAILED = -5
    case inferenceFailed = -5

    /// Invalid context handle. EV_ERROR_CONTEXT_INVALID = -6
    case contextInvalid = -6

    /// Stream has ended (not an error — used as a sentinel). EV_ERROR_STREAM_ENDED = -7
    case streamEnded = -7

    /// Feature not implemented. EV_ERROR_NOT_IMPLEMENTED = -8
    case notImplemented = -8

    /// Memory limit exceeded. EV_ERROR_MEMORY_LIMIT_EXCEEDED = -9
    case memoryLimitExceeded = -9

    /// Backend not supported on this platform. EV_ERROR_UNSUPPORTED_BACKEND = -10
    case unsupportedBackend = -10

    /// Unknown or unmapped error. EV_ERROR_UNKNOWN = -999
    case unknown = -999

    /// Initialize from a raw C `ev_error_t` code, defaulting to `.unknown` for unmapped values.
    public static func from(code: Int32) -> NativeErrorCode {
        return NativeErrorCode(rawValue: code) ?? .unknown
    }

    /// Convert this native error code to an `EdgeVedaError`.
    ///
    /// Returns `nil` for `.ok` and `.streamEnded` since neither represents a failure.
    ///
    /// - Parameter context: Optional additional context string to include in the error message.
    /// - Returns: The corresponding `EdgeVedaError`, or `nil` if the code is not an error.
    public func toEdgeVedaError(context: String? = nil) -> EdgeVedaError? {
        let ctx = context ?? "Native engine error"

        switch self {
        case .ok, .streamEnded:
            return nil

        case .invalidParameter:
            return .invalidParameter(name: "native", value: ctx)

        case .outOfMemory:
            return .outOfMemory

        case .modelLoadFailed:
            return .loadFailed(reason: ctx)

        case .backendInitFailed:
            return .unsupportedBackend(.auto)

        case .inferenceFailed:
            return .generationFailed(reason: ctx)

        case .contextInvalid:
            return .ffiError(message: "Invalid context: \(ctx)")

        case .notImplemented:
            return .ffiError(message: "Not implemented: \(ctx)")

        case .memoryLimitExceeded:
            return .outOfMemory

        case .unsupportedBackend:
            return .unsupportedBackend(.auto)

        case .unknown:
            return .unknown(message: ctx)
        }
    }

    /// Throw the corresponding `EdgeVedaError` if this code is not `.ok`.
    ///
    /// - Parameter context: Optional context string for the error message.
    /// - Throws: `EdgeVedaError` if the code represents an error.
    public func throwIfError(context: String? = nil) throws {
        if let error = toEdgeVedaError(context: context) {
            throw error
        }
    }
}
